### ISSUE TITLE
[feature] override build_requirements in conan install

### ISSUE THREAD

**Issue by gsemet** [conan-io/conan#9302]
This is a follow up of https://github.com/conan-io/conan/issues/2140.

Conan 1.39 introduce this `--require-override` to force a specific version of a package during `conan install`. We might also need the "build_requirements" version of this. Our use case it that we maintain some tools (code generation,...) that have impact on the output files so we might want to test these tools before merging any changes (ex: see if the generated code is still functionnal). We use the same mecanism that what https://github.com/conan-io/conan/issues/2140 is indented into resolving (we override the version of the tools in a Gitlab pipeline). Maybe this is overkill and might be easier to do.

In short, in our recipes, we have one or several package that have "build_requires=sometools/1.2.3@user/stable", we would like to test if "sometool/1.2.4@dev/testing" works or not.

- [ ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @gsemet 

I am checking this feature, and there is an issue that makes it challenging:

> In short, in our recipes, we have one or several package that have "build_requires=sometools/1.2.3@user/stable", we would like to test if "sometool/1.2.4@dev/testing" works or not.

If we map ``--build-require-override="sometool/1.2.4@dev/testing"`` to a profile:
```
[build_requires]
sometool/1.2.4@dev/testing
```
The problem is that it will be injecting that build require to every package in the graph. That might not be the desired behavior, right? This might be alleviated by something like --build-require-override="mypkg:sometool/1.2.4@dev/testing"``, but it would be tedious and error prone to know which packages actually require that tool.

The limitation is that there is no concept of only "overriding" existing recipe ``build_requires``, besides the definition in the profile, but not only when they exists. 

This sounds like a challenging feature to implement, at least with the current graph resolution algorithm. I will explore this concept in Conan 2.0 too when possible.

**Comment by gsemet** [comment]
Our use case is indeed a bit simpler, since it is for a test build, injecting this build_requires for all the graph, and dependencies that would be eventually rebuilded, is the intended behavior.

An analogy is the following one: new version of gcc or cmake is available. Would it make sense to only test it on a single package or on the whole graph ? At the end the graph is "tainted" with this version, and we put a place a small tag in the version (`<version>-ovrd`) in any override cases to avoid any risk of promotion to production (this is for test, so everything should be considered as potentially not safe).

**Comment by Carsten87** [comment]
We are also interested in this feature. For our use case it would also be sufficient to replace the build requirement for the whole graph.

We have some packages we derived from the official (conan-center) ones but with some changes to the recipe (those changes are very specific and it makes no sense to contribute them back to conan-center). To differentiate from the official package we append a `user/channel`, so for example `sometool/1.2.4` becomes `sometool/1.2.4@user/channel`. Now we want to override the build requirement for the whole graph so we don't have to change each recipe which "build requires" `sometool/1.2.4`.

**Comment by slietzau** [comment]
For our use-case a simple `--build-require-override="mypkg:sometool/1.2.4@dev/testing"` would help us solve the issue at least.

To be more specific:
We patch `msys2` and inject our company ssl certificate. To differentiate our version from the official version we publish it as `msys2/latest@company/stable`.  Now if I want to build anything using `msys2` as build_requires, the official version is used. Since `msys2` is a very basic dependency this happens all the time.

We fear that if we just override the official version in our internal artifactory we could break things. However this seems the only sensible solution to this problem. Modifying **all** recipes using `msys2` and adding our `user/channel` seems like a very inconvenient alternative to me.

What would be the conan way to solve this issue? I just stumbled over [alias packages](https://docs.conan.io/en/latest/reference/commands/misc/alias.html). Is that the solution?

**Comment by memsharded** [comment]
@slietzau 
> What would be the conan way to solve this issue? I just stumbled over alias packages. Is that the solution?

No, alias feature is not intended for that case, not recommended here.


For your use case, it seems ``--build-require-override`` will not be the way, but instead to define it in your profile files. As profile files  can be included and concatenated, you could have a specific ``build_requires_profile.txt`` with something like:

```
[build_requires]
mypkg1:msys2/latest@company/stable
mypkg2:msys2/latest@company/stable
```

To have it overriden and injected in the dependencies you want, and not all of them (for all is easy, just avoid the ``mypkg1:`` prefix). If you need extra power, you could make it a ``build_requires_profile.jinja`` and then iterate over a list of packages.

But this feature in particular is for "testing" new potential candidates on the command line without having to modify conanfiles and profiles, but in your case it really seems that you want always to use your ``msys2`` package, and then the recommended approach is to encode that in profile.

**Comment by memsharded** [comment]
We are doing an experiment in next iteration to see if new lockfiles in 2.0 will be a better solution (together with https://github.com/conan-io/conan/pull/10046, same goal)

**Comment by memsharded** [comment]
An update about this issue.
Conan 2.0 has finally settled on the following:
- All the command line ``--override`` arguments have been removed.
- The Conan 2.0 lockfiles allow the ``conan lock add`` command that allow to define new versions into lockfiles. This include ``build_requires`` and ``python_requires`` too.
- As long as the selected versions fit inside declared version ranges in recipes, applying that lockfile will resolve to the versions defined in the lockfile. It is very important that it fits in the version-ranges, lockfiles cannot override outside of those ranges, in that case it would be necessary to go to the recipe and modify the range (such a change also typically means that it is necessary to do other changes to the recipe and code)

Closing the issue as solved in 2.0, thanks for the feedback!

**Comment by gsemet** [comment]
Hi. Does it mean there no way now to test a different version of a dependencies? I agree this makes things complicated, but this important for rapidly identifying regression or do multiple changes at the same time on different packages.

It should generated a 'tainted' package because some dependencies has been overridden, and all packages built on top of this tainted deps should also be tainted. We call these kind of build "engineering build" and should never ends in prod, just for test.

**Comment by memsharded** [comment]
> Hi. Does it mean there no way now to test a different version of a dependencies? I agree this makes things complicated, but this important for rapidly identifying regression or do multiple changes at the same time on different packages.

No, it means the opposite. Now there is a consistent, systematic and traceable way to test different versions of requirements, including normal requires, tool-requires and python-requires.

The only requisite for this is that the specified versions must be valid for the problem definition. That means that they are either revisions (you are testing new revisions), or they fit in the defined version ranges, but this is a very standard thing across package management in general. If a recipe declares that it can use a given ``dep/[>=1.0 <2.0]``, you can test any version in that range easily. But versions outside that range will not be directly overriden, because that exact definition of that range means, "don't even try, a major version upgrade will break, and I will need to adapt my code anyway"

**Comment by gsemet** [comment]
I agree in general this is the wanted behavior, but how would you manage the following use case:

- A depends on B>=1
- B release 2.0 that breaks a important feature.
- A restricts version with B>=1,<2
- B works and finally says it wants to restore the behavior so A can update.

How can A tests unreleased version of B without changing its restriction ?

This is something we do often, let people updates, when something break this is not normal and restriction should always be short living.

What I describe is not the normal "reusable lib" use case where you want your lib to be as much generic as possible. I describe à use case similar to what we can have with Google Repo or Zephyr West: n projects working in parallel in a single "mainline" project and they all have to work with each other. And testing for breakage should be fast and reliable.
Today we trigger pipelines on gitlab with in an environment variable the list of candidate package to test for regression. Most of them are NOT merged yet in their respective master.

**Comment by memsharded** [comment]
I think it is a versioning process problem:

- If the final version of B works without doing changes in A, then it is not a major 2.0, but a minor 1.X
- It is perfectly fine, and recommended to be testing a B 1.X that breaks consumers, until the final decision is made. If we need to change A code, then we will make it a 2.0, if it doesn't need to change A code, then it will be a 1.X version

In all the process, A, doesn't need to change its version range to B.

I am not sure if I understand what you say about google repo or zephyr west (actually I didn't know about the second).
The typical flow for testing new package versions not merged yet in master, is to test against a secondary server repo, where you store the B 1.X versions that you want to test. Only when they are approved and merged to master, they are promoted (copied) to the main server repo, where it will start being used by all developers and builds.

**Comment by gsemet** [comment]
> If the final version of B works without doing changes in A, then it is not a major 2.0, but a minor 1.X

In theory yes, but mistakes happen. semver is hard, and we are all humans. And it also happens a wanted breakage has too major business impact we might choose to revert.

I agree to the "copy to master" repo after merge, but in the repo world, during the heavy development phase, you might want to basically alway fetch the HEAD of all projects, have a CI that verify nothing before merge and merge all at once.

Imagine how a function rename on B would be done:
- "reusable library" behavior: B releases a new feature with a the new function name, keeping the old one for compatibility. It notify about api deprecation. Ultimately in release a breaking change when removing the old function name. This gives a "window" where each consumer can do the necessary adaptation.
- "coherent multi-repo" behavior: all merge/pull requests that rename the function everywhere are opened at the same time on all repositories, and tests are done there is no other breakage. Then merge all at once.

"reusable library" is where package manager such as Conan or npm or poetry are great. "Coherent multi-repo" are where Google repo or West are great, but they do not deal with binaries directly.

What I do is use "Conan" to work in a "coherent multi-repo" environment.

**Comment by memsharded** [comment]
Ok, I understand a bit better now the "coherent multi-repo".

I have some questions about this, if all changes are done simultaneously, then why bother with versions? Versions are a mechanism that is explicit for not having to do all changes at once. If you do all changes at once, then it is not an issue, because you are already modifying the consumers code affected by the change, you can change the version range or whatever is necessary  there. The inconvenience might happen if the consumer code is not affected but you still need to change the version range, because it was not enough to capture the new version. But if your development approach is modifying all affected dependents consumers at once, I say the model fits perfectly:

- When proposing a minor version, in theory consumers don't need to be changed, so the PR change is only done to one package. That specific change can be tested against all the others, just because it automatically fits in the range. It might not even need a lockfile, just putting that new version in the secondary server repo and launching a build against it, will pick it for building.
- When proposing a major version, which is expected that consumers will need to be changed, then it is expected to do 1 PR to the breaking package and several other PRs to the consumers packages that would be breaking because of that change. These consumers packages will modify their ``requirements()`` to account for the new breaking change, as well as their code.
- If the minor version bump fails, because it results to be a major breaking change, then the process enters the major version bump
- If a major version wants to be softened to become a minor, then it is reverting those PRs of the consumers to account for the major.

I'd say this is the totally reasonable and expected behavior, not sure what I could be missing.

**Comment by gsemet** [comment]
The use case you are described is basically what we do AFTER merge.

A few remarks i think worth more reflexion. 

- "When proposing a minor version.." -> mostly agree, this is the easiest change. This is the "normal" workflow once the changes are integrated. But how do you scale this up, when we have dozens of parallel contributions to test? We have a temporary version mecanism to give each PR package a semver version (with the pipeline id in the version for instance), and the "override" mecanism is used to cherry pick only this package downstream. The version range is still respected but there might be several PR to check at the same time, some might work and thus are candidate to be integrated, some might break.
- "When proposing a major version, which is expected that consumers will need to be changed" -> I would rephase to "When proposing a major version, some consumers might need to be changed". This is important, because even a breaking change might not necessarily translate in several MR (ex: unused function renamed for instance). In this case, use has to manually edit their `requirements`... or use the `override` mecanism, which in a sense it the same

Question: how would you set the version range allowing all pre-release just before 2.0.0 (ex: 2.0.0.a01 or 2.0.0.rc3) but it is not in a 1.x ? is there a syntax for the "last of 1.x" ?

**Comment by memsharded** [comment]
This test passes in 2.0:
```python
def test():
    assert Version("2.0.0-pre.1") not in VersionRange(">=1.0 <2")
    assert Version("2.0.0-pre.1") in VersionRange(">=1.0 <2, include_prerelease=True")
    assert Version("2.0.0-pre.1") in VersionRange(">=1.0- <2")
    assert Version("2.0.0-pre.1") in VersionRange(">=1.0 <2-")
```

But the pre-releases should be named as such pre-releases with ``-``, not with ``.`` (``2.0.0.rc1`` is semantically newer than ``2.0.0``, while ``2.0.0-pre.1`` is previous to ``2.0.0``).

Note that in 2.0 the syntax ``-`` is a shortcut for ``include_prerelease=True``.
We have sometimes considered a global conf that allows to opt-in for prereleases at the conf level, but this would not be in 2.0, probably can be considered and discussed in 2.X roadmap.
